### PR TITLE
bpo-36019: Use pythontest.net instead of example.com

### PR DESCRIPTION
**Pull request by matrixise**
https://bugs.python.org/issue36019

### CODE REVIEW

**Comment by vstinner** [comment]
Can you add a NEWS entry?

**Comment by matrixise** [comment]
Sure  but only when my PR won’t be in WIP. Thanks 🙏

**Review by eamanu — APPROVED** [review]
Great!. It would be great use https on the future :-)

**Comment by vstinner** [comment]
> It would be great use https on the future :-)

Wait, some tests must not use HTTPS but HTTP. Some tests are written to explicitly test cleartext HTTP (tcp/80). Right now, it seems like: pythontest.net does't redirect to HTTP.

Maybe we should use a constant to test.support for HTTP-only URL.

**Comment by matrixise** [comment]
@vstinner like that?
```
from test.support import TEST_HTTP_URL
from test.support import TEST_HTTPS_URL
```

and where
```
TEST_HTTP_URL='http://www.pythontest.net'
TEST_HTTP_URL='https://www.pythontest.net'
```

**Comment by vstinner** [comment]
There is no need for HTTPS yet.

**Comment by vstinner** [comment]
"`from test.support import TEST_HTTP_URL" LGTM. Maybe just add a short comment where you define the constant to explain the purpose of this URL. I understood that it must be a public HTTP server, it must use HTTP (not HTTPS), and it must support at least GET verb.

**Comment by matrixise** [comment]
ok, I am going to add this "constant" in `test/support/__init__.py`

File: Lib/test/test_codecmaps_cn.py [hunk 1]
```diff
@@ -4,22 +4,23 @@
...
 
 from test import multibytecodec_support
+from test import support
 import unittest
 
 class TestGB2312Map(multibytecodec_support.TestBase_Mapping,
                    unittest.TestCase):
     encoding = 'gb2312'
-    mapfileurl = 'http://www.pythontest.net/unicode/EUC-CN.TXT'
+    mapfileurl = f'{support.TEST_HTTP_URL}/unicode/EUC-CN.TXT'
```
Comment (vstinner) [inline Lib/test/test_codecmaps_cn.py]: Please revert this change. I would like to be able to modify TEST_HTTP_URL without breaking this URL.

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/test/test_codecmaps_cn.py (hunk 1 above)
Comment (matrixise) [reply Lib/test/test_codecmaps_cn.py]: just this one? or the complete commit?

File: Lib/test/test_hashlib.py [hunk 2]
```diff
@@ -50,7 +50,7 @@ def hexstr(s):
     return r
 
 
-URL = "http://www.pythontest.net/hashlib/{}.txt"
+URL = f"{TEST_HTTP_URL}/hashlib/{{}}.txt"
```
Comment (vstinner) [inline Lib/test/test_hashlib.py]: Don't do that. It will break the test if TEST_HTTP_URL is changed to example.com or anything else.

File: Lib/test/test_urllib.py [hunk 3]
```diff
@@ -712,7 +712,7 @@ def _reporthook(par1, par2, par3):
 
         with self.assertRaises(urllib.error.ContentTooShortError):
             try:
-                urllib.request.urlretrieve('http://example.com/',
+                urllib.request.urlretrieve(f'{support.TEST_HTTP_URL}/',
```
Comment (vstinner) [inline Lib/test/test_urllib.py]: What's the point of adding a slash? Why not using TEST_HTTP_URL directly?
```suggestion
                urllib.request.urlretrieve(support.TEST_HTTP_URL,
```

File: Lib/test/test_urllib2net.py [hunk 4]
```diff
@@ -157,23 +157,23 @@ def test_file(self):
...
         with support.transient_internet(urlwith_frag):
             req = urllib.request.Request(urlwith_frag)
             res = urllib.request.urlopen(req)
             self.assertEqual(res.geturl(),
-                    "http://www.pythontest.net/index.html#frag")
+                    f"{support.TEST_HTTP_URL}/index.html#frag")
 
     def test_redirect_url_withfrag(self):
-        redirect_url_with_frag = "http://www.pythontest.net/redir/with_frag/"
+        redirect_url_with_frag = f"{support.TEST_HTTP_URL}/redir/with_frag/"
```
Comment (vstinner) [inline Lib/test/test_urllib2net.py]: Ditto. Revert such change. The URL becomes wrong if TEST_HTTP_URL becomes example.org, python.org, google.com or whatever else.

File: Doc/library/test.rst [hunk 5]
```diff
@@ -356,6 +356,11 @@ The :mod:`test.support` module defines the following constants:
 
    Check for presence of docstrings.
 
+.. data:: TEST_HTTP_URL
+
+   Define the URL of a dedicated HTTP server for the network tests
+   (http://www.pythontest.net).
```
Comment (vstinner) [inline Doc/library/test.rst]: Please remove the value of the constant, it must not be documented. The value doesn't matter, and it reduces the maintenance burden to not have to update the doc, each time the URL changes.

File: Lib/test/support/__init__.py [hunk 6]
```diff
@@ -835,6 +835,9 @@ def dec(*args, **kwargs):
 # module name.
 TESTFN = "{}_{}_tmp".format(TESTFN, os.getpid())
 
+# Define the URL of a dedicated HTTP server for the network test.
```
Comment (vstinner) [inline Lib/test/support/__init__.py]: Please document that the URL must use clear-text HTTP: no redirection to encrypted HTTPS.

For example, http://www.python.org/ cannot be used: it uses a direction to HTTPS.

**Review by vstinner — CHANGES_REQUESTED** [review]
It seems like you misunderstood what I proposed. The idea of a constant is to adding a constant is to allow to use a different URL without having to update tests. Without your current change, if I modify TEST_HTTP_URL to point to google.com, example.com, etc. A lot of tests fail.

Please restrict the change example.com, and only example.com : https://bugs.python.org/issue36019

File: Lib/test/test_urllibnet.py [hunk 7]
```diff
@@ -24,8 +24,8 @@ def tearDown(self):
         socket.setdefaulttimeout(None)
 
     def testURLread(self):
-        with support.transient_internet("www.example.com"):
-            f = urllib.request.urlopen("http://www.example.com/")
+        with support.transient_internet("www.pythontest.net"):
```
Comment (vstinner) [inline Lib/test/test_urllibnet.py]: Why not using support.TEST_HTTP_URL here?

File: Misc/NEWS.d/next/Tests/2019-02-19-18-30-40.bpo-36019._hBHx7.rst [hunk 8]
```diff
@@ -0,0 +1,2 @@
+Replace http://example.com by http://pythontest.net for some tests.
+Contributed by Stéphane Wirtel
```
Comment (vstinner) [inline Misc/NEWS.d/next/Tests/2019-02-19-18-30-40.bpo-36019._hBHx7.rst]: One NEWS entry is enough. Please merge both NEWS entries.

**Review by vstinner — COMMENTED** [review]
Another request for testURLread().

I blocked connections to example.com using:

```
# I added example.com to localhost
$ sudo head /etc/hosts
127.0.0.1   apu localhost localhost.localdomain localhost4 localhost4.localdomain4 example.com
::1         apu localhost localhost.localdomain localhost6 localhost6.localdomain6 example.com

$ host example.com
example.com has address 93.184.216.34
example.com has IPv6 address 2606:2800:220:1:248:1893:25c8:1946

# Reject all connections to example.com
$ sudo iptables -A OUTPUT -d 93.184.216.34 -j REJECT
$ sudo ip6tables -A OUTPUT -d 2606:2800:220:1:248:1893:25c8:1946 -j REJECT
```

And I modified transient_internet() to get an error:

```
diff --git a/Lib/test/support/__init__.py b/Lib/test/support/__init__.py
index b442070b35..674195e442 100644
--- a/Lib/test/support/__init__.py
+++ b/Lib/test/support/__init__.py
@@ -1483,22 +1483,8 @@ def transient_internet(resource_name, *, timeout=30.0, errnos=()):
     """Return a context manager that raises ResourceDenied when various issues
     with the Internet connection manifest themselves as exceptions."""
     default_errnos = [
-        ('ECONNREFUSED', 111),
-        ('ECONNRESET', 104),
-        ('EHOSTUNREACH', 113),
-        ('ENETUNREACH', 101),
-        ('ETIMEDOUT', 110),
-        # socket.create_connection() fails randomly with
-        # EADDRNOTAVAIL on Travis CI.
-        ('EADDRNOTAVAIL', 99),
     ]
     default_gai_errnos = [
-        ('EAI_AGAIN', -3),
-        ('EAI_FAIL', -4),
-        ('EAI_NONAME', -2),
-        ('EAI_NODATA', -5),
-        # Encountered when trying to resolve IPv6-only hostnames
-        ('WSANO_DATA', 11004),
     ]
 
     denied = ResourceDenied("Resource %r is not available" % resource_name)
```

Only test_timeout failed because of my patch. So it's good.

File: Lib/test/test_urllibnet.py (hunk 7 above)
Comment (matrixise) [reply Lib/test/test_urllibnet.py]: because `support.transient_internet` uses a hostname and not scheme+hostname.

Comment (matrixise) [reply Lib/test/test_urllibnet.py]: for that, I have introduced `support.TEST_HOSTNAME`

File: Lib/test/test_urllibnet.py [hunk 9]
```diff
@@ -24,8 +24,8 @@ def tearDown(self):
         socket.setdefaulttimeout(None)
 
     def testURLread(self):
-        with support.transient_internet("www.example.com"):
-            f = urllib.request.urlopen("http://www.example.com/")
+        with support.transient_internet(support.TEST_HOSTNAME):
```
Comment (vstinner) [inline Lib/test/test_urllibnet.py]: Would it be possible to extract the hostname from the URL rather than introducing another constant? It would avoid the risk of avoid 2 constants which would be inconsistent (if 1 is updated but not the other).

urllib.parse has tools for that.

Comment (matrixise) [reply Lib/test/test_urllibnet.py]: yep, we can do it. I am going to fix it.

**Comment by matrixise** [comment]
@vstinner updated, when you have time.

File: Lib/test/test_urllibnet.py [hunk 10]
```diff
@@ -24,8 +25,9 @@ def tearDown(self):
         socket.setdefaulttimeout(None)
 
     def testURLread(self):
-        with support.transient_internet("www.example.com"):
-            f = urllib.request.urlopen("http://www.example.com/")
+        domain = urllib.parse.urlparse(support.TEST_HTTP_URL).netloc
```
Comment (vstinner) [inline Lib/test/test_urllibnet.py:28]: Good :-)

**Review by vstinner — APPROVED** [review]
LGTM.
